File: slot.js
// slot.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');

const COLORS = {
    reset: '\x1b[0m',
    red: '\x1b[91m',
    green: '\x1b[92m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    magenta: '\x1b[95m',
    cyan: '\x1b[96m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const SYMBOLS = ['🍒', '🍋', '🍊', '🍇', '🔔', '💎', '7'];
const SYMBOL_COLORS = ['red', 'yellow', 'magenta', 'magenta', 'cyan', 'cyan', 'bold'];
const SYMBOL_VALUES = {
    '🍒': 5, '🍋': 5, '🍊': 5, '🍇': 8, '🔔': 10, '💎': 100, '7': 50
};

class SlotMachine {
    constructor() {
        this.balance = 100;
        this.bet = 0;
        this.history = [];
        this.statsFile = path.join(os.homedir(), '.slot_stats.json');
        this.loadStats();
    }

    loadStats() {
        try {
            const data = JSON.parse(fs.readFileSync(this.statsFile, 'utf8'));
            this.rolls = data.rolls || 0;
            this.totalWon = data.total_won || 0;
            this.bestWin = data.best_win || 0;
            this.balance = data.balance || 100;
        } catch {
            this.rolls = 0;
            this.totalWon = 0;
            this.bestWin = 0;
            this.balance = 100;
        }
    }

    saveStats() {
        const data = {
            rolls: this.rolls,
            total_won: this.totalWon,
            best_win: this.bestWin,
            balance: this.balance
        };
        fs.writeFileSync(this.statsFile, JSON.stringify(data, null, 2));
    }

    spin(bet) {
        if (bet > this.balance) {
            return { result: null, msg: 'Недостаточно средств.' };
        }
        if (bet <= 0) {
            return { result: null, msg: 'Ставка должна быть положительной.' };
        }
        this.bet = bet;
        this.balance -= bet;
        // Анимация
        console.log(colorize('Вращение...', 'bold'));
        for (let i = 0; i < 3; i++) {
            let temp = SYMBOLS.map(() => SYMBOLS[Math.floor(Math.random() * SYMBOLS.length)]);
            process.stdout.write(`\r${temp[0]} ${temp[1]} ${temp[2]}`);
            // Задержка
            const start = Date.now();
            while (Date.now() - start < 200) {}
        }
        // Результат
        const result = SYMBOLS.map(() => SYMBOLS[Math.floor(Math.random() * SYMBOLS.length)]);
        process.stdout.write(`\r${result[0]} ${result[1]} ${result[2]}\n`);
        const win = this.calculateWin(result);
        this.balance += win;
        this.rolls++;
        this.totalWon += win;
        if (win > this.bestWin) this.bestWin = win;
        this.history.push(result);
        if (win > 0) {
            console.log(colorize(`Вы выиграли ${win} монет!`, 'green'));
        } else {
            console.log(colorize('Ничего не выиграно.', 'red'));
        }
        this.saveStats();
        return { result, win };
    }

    calculateWin(result) {
        if (result[0] === '💎' && result[1] === '💎' && result[2] === '💎') {
            return this.bet * 100;
        }
        if (result[0] === result[1] && result[1] === result[2]) {
            return this.bet * (SYMBOL_VALUES[result[0]] || 1);
        }
        // Два одинаковых + джокер
        for (const sym of SYMBOLS) {
            if (sym === '💎') continue;
            const count = result.filter(s => s === sym).length;
            if (count === 2 && result.includes('💎')) {
                return this.bet * 2 * SYMBOL_VALUES[sym];
            }
        }
        if (result.includes('💎')) {
            return this.bet * 2;
        }
        return 0;
    }

    displayBalance() {
        console.log(colorize(`Баланс: ${this.balance} монет`, 'yellow'));
    }

    displayStats() {
        console.log(colorize('📊 Статистика:', 'bold'));
        console.log(`  Вращений: ${this.rolls}`);
        console.log(`  Всего выиграно: ${this.totalWon}`);
        console.log(`  Лучший выигрыш: ${this.bestWin}`);
        console.log(`  Текущий баланс: ${this.balance}`);
    }

    async run() {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        const question = (q) => new Promise(resolve => rl.question(q, resolve));

        console.log(colorize('🎰 Добро пожаловать в Слот-машину!', 'bold'));
        console.log('Команды: spin <ставка>, balance, stats, deposit <сумма>, help, quit');
        this.displayBalance();

        while (true) {
            const input = await question('> ');
            const parts = input.trim().split(/\s+/);
            const cmd = parts[0]?.toLowerCase();
            if (!cmd) continue;
            if (cmd === 'quit' || cmd === 'q') {
                console.log('Выход.');
                this.saveStats();
                rl.close();
                break;
            } else if (cmd === 'balance' || cmd === 'b') {
                this.displayBalance();
            } else if (cmd === 'stats') {
                this.displayStats();
            } else if (cmd === 'spin') {
                if (parts.length !== 2) {
                    console.log('Используйте: spin <ставка>');
                    continue;
                }
                const bet = parseInt(parts[1]);
                if (isNaN(bet)) {
                    console.log('Ставка должна быть числом.');
                    continue;
                }
                const result = this.spin(bet);
                if (result.result) {
                    // уже выведено
                } else {
                    console.log(colorize(result.msg, 'red'));
                }
            } else if (cmd === 'deposit') {
                if (parts.length !== 2) {
                    console.log('Используйте: deposit <сумма>');
                    continue;
                }
                const amount = parseInt(parts[1]);
                if (isNaN(amount) || amount <= 0) {
                    console.log('Сумма должна быть положительным числом.');
                    continue;
                }
                this.balance += amount;
                console.log(colorize(`Баланс пополнен на ${amount}. Текущий баланс: ${this.balance}`, 'green'));
                this.saveStats();
            } else if (cmd === 'help' || cmd === 'h') {
                console.log('Команды:');
                console.log('  spin <ставка>  - сделать ставку и крутить барабаны');
                console.log('  balance        - показать баланс');
                console.log('  stats          - показать статистику');
                console.log('  deposit <сумма> - пополнить баланс');
                console.log('  quit           - выход');
            } else {
                console.log('Неизвестная команда. Введите \'help\' для справки.');
            }
        }
    }
}

async function main() {
    const game = new SlotMachine();
    await game.run();
}

main().catch(console.error);

File: slot.go
// slot.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	red    = "\033[91m"
	green  = "\033[92m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	magenta= "\033[95m"
	cyan   = "\033[96m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

var symbols = []string{"🍒", "🍋", "🍊", "🍇", "🔔", "💎", "7"}
var symbolColors = []string{red, yellow, magenta, magenta, cyan, cyan, bold}
var symbolValues = map[string]int{
	"🍒": 5, "🍋": 5, "🍊": 5, "🍇": 8, "🔔": 10, "💎": 100, "7": 50,
}

type Stats struct {
	Rolls    int `json:"rolls"`
	TotalWon int `json:"total_won"`
	BestWin  int `json:"best_win"`
	Balance  int `json:"balance"`
}

type SlotMachine struct {
	balance  int
	bet      int
	history  [][]string
	stats    Stats
	statsFile string
}

func NewSlotMachine() *SlotMachine {
	s := &SlotMachine{balance: 100}
	s.statsFile = filepath.Join(os.Getenv("HOME"), ".slot_stats.json")
	s.loadStats()
	return s
}

func (s *SlotMachine) loadStats() {
	data, err := os.ReadFile(s.statsFile)
	if err != nil {
		s.stats = Stats{Balance: 100}
		return
	}
	if err := json.Unmarshal(data, &s.stats); err != nil {
		s.stats = Stats{Balance: 100}
	}
	if s.stats.Balance == 0 && s.stats.Rolls == 0 {
		s.stats.Balance = 100
	}
	s.balance = s.stats.Balance
}

func (s *SlotMachine) saveStats() {
	s.stats.Balance = s.balance
	data, _ := json.MarshalIndent(s.stats, "", "  ")
	os.WriteFile(s.statsFile, data, 0644)
}

func (s *SlotMachine) spin(bet int) []string {
	if bet > s.balance {
		fmt.Println(colorize("Недостаточно средств.", red))
		return nil
	}
	if bet <= 0 {
		fmt.Println(colorize("Ставка должна быть положительной.", red))
		return nil
	}
	s.bet = bet
	s.balance -= bet
	fmt.Println(colorize("Вращение...", bold))
	// Анимация
	for i := 0; i < 3; i++ {
		fmt.Print("\r")
		for j := 0; j < 3; j++ {
			idx := rand.Intn(len(symbols))
			fmt.Print(colorize(symbols[idx], symbolColors[idx]) + " ")
		}
		time.Sleep(200 * time.Millisecond)
	}
	// Результат
	result := make([]string, 3)
	for i := 0; i < 3; i++ {
		idx := rand.Intn(len(symbols))
		result[i] = symbols[idx]
	}
	fmt.Printf("\r%s %s %s\n",
		colorize(result[0], symbolColors[indexOf(symbols, result[0])]),
		colorize(result[1], symbolColors[indexOf(symbols, result[1])]),
		colorize(result[2], symbolColors[indexOf(symbols, result[2])]),
	)
	win := s.calculateWin(result)
	s.balance += win
	s.stats.Rolls++
	s.stats.TotalWon += win
	if win > s.stats.BestWin {
		s.stats.BestWin = win
	}
	s.history = append(s.history, result)
	if win > 0 {
		fmt.Printf("%s\n", colorize(fmt.Sprintf("Вы выиграли %d монет!", win), green))
	} else {
		fmt.Println(colorize("Ничего не выиграно.", red))
	}
	s.saveStats()
	return result
}

func indexOf(slice []string, val string) int {
	for i, v := range slice {
		if v == val {
			return i
		}
	}
	return 0
}

func (s *SlotMachine) calculateWin(result []string) int {
	if result[0] == "💎" && result[1] == "💎" && result[2] == "💎" {
		return s.bet * 100
	}
	if result[0] == result[1] && result[1] == result[2] {
		return s.bet * symbolValues[result[0]]
	}
	// Два одинаковых + джокер
	for _, sym := range symbols {
		if sym == "💎" {
			continue
		}
		count := 0
		for _, r := range result {
			if r == sym {
				count++
			}
		}
		if count == 2 && contains(result, "💎") {
			return s.bet * 2 * symbolValues[sym]
		}
	}
	if contains(result, "💎") {
		return s.bet * 2
	}
	return 0
}

func contains(slice []string, val string) bool {
	for _, v := range slice {
		if v == val {
			return true
		}
	}
	return false
}

func (s *SlotMachine) displayBalance() {
	fmt.Printf("%s\n", colorize(fmt.Sprintf("Баланс: %d монет", s.balance), yellow))
}

func (s *SlotMachine) displayStats() {
	fmt.Println(colorize("📊 Статистика:", bold))
	fmt.Printf("  Вращений: %d\n", s.stats.Rolls)
	fmt.Printf("  Всего выиграно: %d\n", s.stats.TotalWon)
	fmt.Printf("  Лучший выигрыш: %d\n", s.stats.BestWin)
	fmt.Printf("  Текущий баланс: %d\n", s.balance)
}

func (s *SlotMachine) run() {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println(colorize("🎰 Добро пожаловать в Слот-машину!", bold))
	fmt.Println("Команды: spin <ставка>, balance, stats, deposit <сумма>, help, quit")
	s.displayBalance()
	for {
		fmt.Print("> ")
		if !scanner.Scan() {
			break
		}
		cmd := strings.TrimSpace(scanner.Text())
		parts := strings.Fields(cmd)
		if len(parts) == 0 {
			continue
		}
		switch parts[0] {
		case "quit", "q":
			fmt.Println("Выход.")
			s.saveStats()
			return
		case "balance", "b":
			s.displayBalance()
		case "stats":
			s.displayStats()
		case "spin":
			if len(parts) != 2 {
				fmt.Println("Используйте: spin <ставка>")
				continue
			}
			bet, err := strconv.Atoi(parts[1])
			if err != nil {
				fmt.Println("Ставка должна быть числом.")
				continue
			}
			s.spin(bet)
		case "deposit":
			if len(parts) != 2 {
				fmt.Println("Используйте: deposit <сумма>")
				continue
			}
			amount, err := strconv.Atoi(parts[1])
			if err != nil || amount <= 0 {
				fmt.Println("Сумма должна быть положительным числом.")
				continue
			}
			s.balance += amount
			fmt.Printf("%s\n", colorize(fmt.Sprintf("Баланс пополнен на %d. Текущий баланс: %d", amount, s.balance), green))
			s.saveStats()
		case "help", "h":
			fmt.Println("Команды:")
			fmt.Println("  spin <ставка>  - сделать ставку и крутить барабаны")
			fmt.Println("  balance        - показать баланс")
			fmt.Println("  stats          - показать статистику")
			fmt.Println("  deposit <сумма> - пополнить баланс")
			fmt.Println("  quit           - выход")
		default:
			fmt.Println("Неизвестная команда. Введите 'help' для справки.")
		}
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())
	game := NewSlotMachine()
	game.run()
}
